// @syncano/core/lib/endpoint.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _queryBuilder = require('./query-builder');

var _queryBuilder2 = _interopRequireDefault(_queryBuilder);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Running endpoints.
 * @property {Function}
 * @example {@lang javascript}
 * const latestTags = await socket.get('tags/list', { sort: 'latest' })
 * const createdTag = await socket.post('tags/create', { name: 'nature' })
 */
class Endpoint extends _queryBuilder2.default {
  post(endpoint, body = {}, options = {}) {
    const fetch = this.fetch.bind(this);

    return fetch(this._url(endpoint), _extends({
      method: 'POST',
      body: this._parseBody(body)
    }, options));
  }

  get(endpoint, data = {}, options = {}) {
    return this.post(endpoint, _extends({}, data, { _method: 'GET' }), options);
  }

  delete(endpoint, data = {}, options = {}) {
    return this.post(endpoint, _extends({}, data, { _method: 'DELETE' }), options);
  }

  put(endpoint, data = {}, options = {}) {
    return this.post(endpoint, _extends({}, data, { _method: 'PUT' }), options);
  }

  patch(endpoint, data = {}, options = {}) {
    return this.post(endpoint, _extends({}, data, { _method: 'PATCH' }), options);
  }

  _url(endpoint) {
    const { instanceName, spaceHost } = this.instance;

    return `https://${instanceName}.${spaceHost}/${endpoint}/`;
  }

  _parseBody(body) {
    const isBodyAnObject = typeof body === 'object';

    return isBodyAnObject ? JSON.stringify(_extends({}, body)) : body;
  }
}
exports.default = Endpoint;
module.exports = exports['default'];
