// @syncano/cli/lib/settings/accountSettings.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _settings = require('./settings');

var _settings2 = _interopRequireDefault(_settings);

var _debug = require('../utils/debug');

var _debug2 = _interopRequireDefault(_debug);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const { debug } = (0, _debug2.default)('settings-account');

class AccountSettings extends _settings2.default {
  constructor() {
    super();
    this.name = process.env.SYNCANO_ACCOUNT_FILE ? process.env.SYNCANO_ACCOUNT_FILE : 'syncano';
    this.baseDir = _os2.default.homedir();
    this.loaded = this.load();
  }

  logout() {
    debug('logout()');
    delete this.attributes.auth_key;
    this.save();
  }

  authenticated() {
    return Boolean(this.getAuthKey());
  }

  getAuthKey() {
    return process.env.SYNCANO_AUTH_KEY || this.attributes.auth_key || null;
  }

  addProject(projectName, params = {}) {
    debug('addProject()', projectName);
    if (!this.attributes.projects) {
      this.attributes.projects = {};
    }
    this.attributes.projects[projectName] = params;
    this.save();
  }

  getProject(projectName) {
    debug('getProject()', projectName);
    const envInstance = process.env.SYNCANO_PROJECT_INSTANCE;
    if (envInstance) {
      return { instance: envInstance };
    }
    if (this.attributes.projects) {
      return this.attributes.projects[projectName];
    }
  }
}
exports.default = AccountSettings;
module.exports = exports['default'];
