// @syncano/cli/lib/commands/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _attach = require('./attach');

var _attach2 = _interopRequireDefault(_attach);

var _init = require('./init');

var _init2 = _interopRequireDefault(_init);

var _login = require('./login');

var _login2 = _interopRequireDefault(_login);

var _logout = require('./logout');

var _logout2 = _interopRequireDefault(_logout);

var _hostingAdd = require('./hosting-add');

var _hostingAdd2 = _interopRequireDefault(_hostingAdd);

var _hostingDelete = require('./hosting-delete');

var _hostingDelete2 = _interopRequireDefault(_hostingDelete);

var _hostingList = require('./hosting-list');

var _hostingList2 = _interopRequireDefault(_hostingList);

var _hostingFiles = require('./hosting-files');

var _hostingFiles2 = _interopRequireDefault(_hostingFiles);

var _hostingSync = require('./hosting-sync');

var _hostingSync2 = _interopRequireDefault(_hostingSync);

var _hostingConfig = require('./hosting-config');

var _hostingConfig2 = _interopRequireDefault(_hostingConfig);

var _socketList = require('./socket-list');

var _socketList2 = _interopRequireDefault(_socketList);

var _socketDeploy = require('./socket-deploy');

var _socketDeploy2 = _interopRequireDefault(_socketDeploy);

var _socketDeployHot = require('./socket-deploy-hot');

var _socketDeployHot2 = _interopRequireDefault(_socketDeployHot);

var _socketConfig = require('./socket-config');

var _socketConfig2 = _interopRequireDefault(_socketConfig);

var _socketConfigSet = require('./socket-config-set');

var _socketConfigSet2 = _interopRequireDefault(_socketConfigSet);

var _socketConfigShow = require('./socket-config-show');

var _socketConfigShow2 = _interopRequireDefault(_socketConfigShow);

var _socketCall = require('./socket-call');

var _socketCall2 = _interopRequireDefault(_socketCall);

var _socketSearch = require('./socket-search');

var _socketSearch2 = _interopRequireDefault(_socketSearch);

var _socketSubmit = require('./socket-submit');

var _socketSubmit2 = _interopRequireDefault(_socketSubmit);

var _socketPublish = require('./socket-publish');

var _socketPublish2 = _interopRequireDefault(_socketPublish);

var _socketTrace = require('./socket-trace');

var _socketTrace2 = _interopRequireDefault(_socketTrace);

var _socketInstall = require('./socket-install');

var _socketInstall2 = _interopRequireDefault(_socketInstall);

var _socketCreate = require('./socket-create');

var _socketCreate2 = _interopRequireDefault(_socketCreate);

var _socketUninstall = require('./socket-uninstall');

var _socketUninstall2 = _interopRequireDefault(_socketUninstall);

var _componentList = require('./component-list');

var _componentList2 = _interopRequireDefault(_componentList);

var _componentLink = require('./component-link');

var _componentLink2 = _interopRequireDefault(_componentLink);

var _instanceList = require('./instance-list');

var _instanceList2 = _interopRequireDefault(_instanceList);

var _instanceCreate = require('./instance-create');

var _instanceCreate2 = _interopRequireDefault(_instanceCreate);

var _instanceDelete = require('./instance-delete');

var _instanceDelete2 = _interopRequireDefault(_instanceDelete);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  Attach: _attach2.default,
  Init: _init2.default,
  Login: _login2.default,
  Logout: _logout2.default,
  HostingAdd: _hostingAdd2.default,
  HostingDelete: _hostingDelete2.default,
  HostingList: _hostingList2.default,
  HostingFilesCmd: _hostingFiles2.default,
  HostingSync: _hostingSync2.default,
  SocketDeployHot: _socketDeployHot2.default,
  HostingConfig: _hostingConfig2.default,
  SocketDeploy: _socketDeploy2.default,
  SocketConfig: _socketConfig2.default,
  SocketConfigSet: _socketConfigSet2.default,
  SocketConfigShow: _socketConfigShow2.default,
  SocketList: _socketList2.default,
  SocketEndpointCall: _socketCall2.default,
  SocketSearch: _socketSearch2.default,
  SocketPublish: _socketPublish2.default,
  SocketSubmit: _socketSubmit2.default,
  SocketTrace: _socketTrace2.default,
  SocketInstall: _socketInstall2.default,
  SocketCreate: _socketCreate2.default,
  SocketUninstall: _socketUninstall2.default,
  ComponentList: _componentList2.default,
  ComponentLink: _componentLink2.default,
  InstanceList: _instanceList2.default,
  InstanceCreate: _instanceCreate2.default,
  InstanceDelete: _instanceDelete2.default
};
module.exports = exports['default'];
